app/services/encryption/encryptors/attribute_encryptor.rb

```diff
@@ -0,0 +1,67 @@
+module Encryption
+  module Encryptors
+    class AttributeEncryptor
+      def encrypt(plaintext)
+        user_access_key = self.class.load_or_init_user_access_key(
+          key: current_key, cost: current_cost
+        )
+        UserAccessKeyEncryptor.new(user_access_key).encrypt(plaintext)
+      end
+
+      def decrypt(ciphertext)
+        encryption_keys_with_cost.each do |key_with_cost|
+          key = key_with_cost.fetch(:key)
+          cost = key_with_cost.fetch(:cost)
+          result = try_decrypt(ciphertext, key: key, cost: cost)
+          return result unless result.nil?
+        end
+        raise Pii::EncryptionError, 'unable to decrypt attribute with any key'
+      end
+
+      def stale?
+        return false if stale.nil?
+        stale
+      end
+
+      def self.load_or_init_user_access_key(key:, cost:)
+        @_scypt_hashes_by_key ||= {}
+        scrypt_hash = @_scypt_hashes_by_key["#{key}:#{cost}"]
+        return UserAccessKey.new(scrypt_hash: scrypt_hash) if scrypt_hash.present?
+        uak = UserAccessKey.new(password: key, salt: key, cost: cost)
+        @_scypt_hashes_by_key["#{key}:#{cost}"] = uak.as_scrypt_hash
+        uak
+      end
+
+      private
+
+      attr_accessor :stale
+
+      def try_decrypt(ciphertext, key:, cost:)
+        user_access_key = self.class.load_or_init_user_access_key(key: key, cost: cost)
+        begin
+          result = UserAccessKeyEncryptor.new(user_access_key).decrypt(ciphertext)
+          self.stale = true if key != current_key
```

Comment:
if `self.stale` was `true` and `key == current_key`, is it still stale? I'm not seeing anything that would make that the case, so perhaps `self.stale = key != current_key`?

Reply:
Ah, nice. Yeah, that works.

Reply:
Ultimately, I'd like to find a way to pull this out, because I don't want to much of the user model's implementation to inform how this works, but we'll get there.

Reply:
Okay, I've clean this on up. Lmk if you see anything else.